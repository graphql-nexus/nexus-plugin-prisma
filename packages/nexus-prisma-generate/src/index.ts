#!/usr/bin/env node

import { existsSync, mkdirSync, writeFileSync } from 'fs'
import {
  GraphQLEnumType,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLSchema,
  isEnumType,
  isInputObjectType,
  isObjectType,
  isScalarType,
  introspectionFromSchema,
} from 'graphql'
import * as meow from 'meow'
import { EOL } from 'os'
import { join } from 'path'
import { ISDL } from 'prisma-datamodel'
import { generateCRUDSchemaFromInternalISDL } from 'prisma-generate-schema'
import {
  findDatamodelAndComputeSchema,
  findRootDirectory,
  getImportPathRelativeToOutput,
} from './config'
import { getFinalType, getTypeName, isList, isRequired } from './graphql'

const cli = meow(
  `
    nexus-prisma-generate prisma-client-dir output

    > Generate the TypeScript types for nexus-prisma

    -----
    
    Inputs should be relative to the root of your project

    --client (required): Path to your prisma-client directory (eg: ./generated/prisma-client/)
    --output (required): Path to directory where you want to output the typings (eg: ./generated/nexus-prisma)
    --js     (optional): Whether to generate the types for Javascript
`,
  {
    flags: {
      client: {
        type: 'string',
      },
      output: {
        type: 'string',
      },
      js: {
        type: 'boolean',
        default: false,
      },
    },
  },
)

main(cli)

function main(cli: meow.Result) {
  const { client: prismaClientDir, output, js: jsMode } = cli.flags

  if (!prismaClientDir) {
    console.log('ERROR: Missing argument --client')
    process.exit(1)
  }

  if (!output) {
    console.log('ERROR: Missing argument --output')
    process.exit(1)
  }

  const rootPath = findRootDirectory()
  const resolvedOutput = output.startsWith('/')
    ? output
    : join(rootPath, output)
  const resolvedPrismaClientDir = prismaClientDir.startsWith('/')
    ? prismaClientDir
    : join(rootPath, prismaClientDir)

  if (!existsSync(resolvedPrismaClientDir)) {
    console.log(
      `ERROR: Cannot find --client path at ${resolvedPrismaClientDir}`,
    )
    process.exit(1)
  }

  try {
    // Create the output directories if needed (mkdir -p)
    mkdirSync(resolvedOutput, { recursive: true })
  } catch (e) {
    if (e.code !== 'EEXIST') throw e
  }

  const { datamodel, databaseType } = findDatamodelAndComputeSchema()

  try {
    const schema = generateCRUDSchemaFromInternalISDL(datamodel, databaseType)
    const renderedDatamodel = renderDatamodel(datamodel, schema)
    const nexusPrismaTypesPath = join(rootPath, output, 'nexus-prisma.ts')
    const nexusPrismaTypes = renderNexusPrismaTypes(
      schema,
      getImportPathRelativeToOutput(
        resolvedPrismaClientDir,
        nexusPrismaTypesPath,
      ),
    )

    writeFileSync(nexusPrismaTypesPath, nexusPrismaTypes)

    if (jsMode) {
      const datamodelPath = join(rootPath, output, 'datamodel-info.js')
      const indexPath = join(rootPath, output, 'index.js')

      writeFileSync(datamodelPath, `module.exports = ${renderedDatamodel}`)
      writeFileSync(indexPath, renderIndexJs())
    } else {
      const datamodelPath = join(rootPath, output, 'datamodel-info.ts')
      const indexPath = join(rootPath, output, 'index.ts')

      writeFileSync(datamodelPath, `export default ${renderedDatamodel}`)
      writeFileSync(indexPath, `export { default } from './datamodel-info'`)
    }

    console.log(`Types generated at ${output}`)
  } catch (e) {
    console.error(e)
  }
}

function renderIndexJs() {
  return `\
const datamodelInfo = require('./datamodel-info')
  
module.exports = datamodelInfo
  `
}

function renderDatamodel(datamodel: ISDL, schema: GraphQLSchema) {
  return `\
{
  uniqueFieldsByModel: {
${datamodel.types
  .map(
    type =>
      `    ${type.name}: [${type.fields
        .filter(field => field.isUnique)
        .map(field => `'${field.name}'`)
        .join(', ')}]`,
  )
  .join(',' + EOL)}
  },
  schema: ${JSON.stringify(introspectionFromSchema(schema), null, 2)}
}
  `
}

function renderNexusPrismaTypes(
  schema: GraphQLSchema,
  prismaClientPath: string,
) {
  const types = Object.values(schema.getTypeMap())
  const objectTypes = types.filter(
    t => isObjectType(t) && !t.name.startsWith('__'),
  ) as GraphQLObjectType[]
  const inputTypes = types.filter(isInputObjectType)
  const enumTypes = types.filter(
    t => isEnumType(t) && !t.name.startsWith('__'),
  ) as GraphQLEnumType[]

  return `\
// GENERATED TYPES FOR NEXUS-PRISMA. /!\\ DO NOT EDIT MANUALLY

import { core } from 'nexus'
import { GraphQLResolveInfo } from 'graphql'
import * as prisma from '${prismaClientPath}'

declare global {
  interface NexusPrismaGen extends NexusPrismaTypes {}
}

export interface NexusPrismaTypes {
  objectTypes: {
    fields: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsName(type)}`)
  .join(EOL)}
    }
    fieldsDetails: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsDetailsName(type)}`)
  .join(EOL)}
    }
  }
  inputTypes: {
    fields: {
${inputTypes
  .map(
    type =>
      `      ${getInputObjectTypeName(type)}: ${getInputObjectTypeFieldsName(
        type,
      )}`,
  )
  .join(EOL)}
    }
  }
  enumTypes: {
${enumTypes
  .map(type => `    ${type.name}: ${getEnumTypeName(type)},`)
  .join(EOL)}
  }
}

${objectTypes.map(renderObjectType).join(EOL)}

${inputTypes.map(renderInputType).join(EOL)}

${enumTypes.map(renderEnumType).join(EOL)}
  `
}

function renderObjectType(type: GraphQLObjectType) {
  const fields = Object.values(type.getFields())

  return `\
// Types for ${type.name}

${renderFields(type, fields)}

${renderFieldsArgs(type, fields)}

${renderTypeFieldDetails(type, fields)}
`
}

function renderTypeFieldDetails(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
export interface ${getObjectTypeFieldsDetailsName(type)} {
${fields
  .map(
    field => `\
  ${field.name}: {
    type: '${getTypeName(field.type)}'
    args: ${
      field.args.length > 0
        ? `Record<${getTypeFieldArgName(
            type,
            field,
          )}, core.NexusArgDef<string>>`
        : '{}'
    }
    description: string
    list: ${isList(field.type) ? true : undefined}
    nullable: ${!isRequired(field.type)}
    resolve: ${
      isScalarType(getFinalType(field.type))
        ? undefined
        : `(
      root: core.RootValue<"${type.name}">,
      args: ${renderResolverArgs(field)},
      context: core.GetGen<"context">,
      info?: GraphQLResolveInfo
    ) => ${renderResolverReturnType(field)}`
    }
  }`,
  )
  .join(EOL)}
}
  `
}

function renderResolverArgs(field: GraphQLField<any, any>) {
  return `{\
 ${field.args
   .map(
     arg => `${arg.name}${isRequired(arg.type) ? '' : '?'}: ${getTSType(arg)}`,
   )
   .join(', ')} }\
  `
}

function renderFields(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
type ${getObjectTypeFieldsName(type)} =
  | ${getExposableFieldsTypeName(type)}
${fields
  .map(
    f =>
      `  | { name: '${f.name}', args?: ${
        f.args.length > 0
          ? `${getTypeFieldArgName(type, f)}[] | false`
          : '[] | false'
      }, alias?: string  } `,
  )
  .join(EOL)}

type ${getExposableFieldsTypeName(type)} =
${fields.map(f => `  | '${f.name}'`).join(EOL)}
`
}

function renderFieldsArgs(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
${fields
  .filter(field => field.args.length > 0)
  .map(field => renderFieldArg(type, field))
  .join(EOL)}
  `
}

function renderFieldArg(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `\
type ${getTypeFieldArgName(type, field)} =
${field.args.map(arg => `  | '${arg.name}'`).join(EOL)}`
}

function getTSType(graphqlType: GraphQLField<any, any> | GraphQLInputField) {
  const graphqlToTypescript: Record<string, string> = {
    String: 'string',
    Boolean: 'boolean',
    ID: 'string',
    Int: 'number',
    Float: 'number',
    DateTime: 'string',
  }
  const finalType = getFinalType(graphqlType.type)

  let returnType = ''

  if (isScalarType(finalType)) {
    returnType = graphqlToTypescript[getTypeName(finalType)]
  } else if (isInputObjectType(finalType)) {
    returnType = getInputObjectTypeName(finalType)
  } else {
    returnType = `prisma.${getTypeName(finalType)}`
  }

  if (isList(graphqlType.type)) {
    returnType += '[]'
  }

  if (!isRequired(graphqlType.type)) {
    returnType += ' | null'
  }

  return returnType
}

function renderResolverReturnType(field: GraphQLField<any, any>) {
  const returnType = getTSType(field)

  return `Promise<${returnType}> | ${returnType}`
}

function renderInputType(input: GraphQLInputObjectType) {
  const fields = Object.values(input.getFields())
  return `\
export interface ${getInputObjectTypeName(input)} {
${fields
  .map(
    field =>
      `  ${field.name}${isRequired(input) ? '' : '?'}: ${getTSType(field)}`,
  )
  .join(EOL)}
}
export type ${getInputObjectTypeFieldsName(input)} =
  | Extract<keyof ${getInputObjectTypeName(input)}, string>
${fields.map(f => `  | { name: '${f.name}', alias?: string  } `).join(EOL)}
  `
}

function renderEnumType(enumType: GraphQLEnumType): string {
  return `\
export type ${getEnumTypeName(enumType)} =
${enumType
  .getValues()
  .map(value => `  | '${value.name}'`)
  .join(EOL)}
  `
}

function getExposableFieldsTypeName(type: GraphQLObjectType) {
  return `${type.name}Fields`
}

function upperFirst(s: string) {
  return s.replace(/^\w/, c => c.toUpperCase())
}

function getTypeFieldArgName(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `${type.name}${upperFirst(field.name)}Args`
}

function getObjectTypeFieldsName(type: GraphQLObjectType) {
  return `${type.name}Object`
}

function getInputObjectTypeFieldsName(type: GraphQLInputObjectType) {
  return `${type.name}InputObject`
}

function getObjectTypeFieldsDetailsName(type: GraphQLObjectType) {
  return `${type.name}FieldDetails`
}

function getInputObjectTypeName(type: GraphQLInputObjectType) {
  return `${type.name}`
}

function getEnumTypeName(enumType: GraphQLEnumType): string {
  return `${enumType.name}Values`
}
