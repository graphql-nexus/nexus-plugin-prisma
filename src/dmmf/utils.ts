import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
