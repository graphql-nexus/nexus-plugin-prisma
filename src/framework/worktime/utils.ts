import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
