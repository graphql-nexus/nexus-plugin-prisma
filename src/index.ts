import * as Builder from './builder'

export type NexusPrismaParams = Builder.Options
export const nexusPrismaPlugin = Builder.build
