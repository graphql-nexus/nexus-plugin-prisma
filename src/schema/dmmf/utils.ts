import { DMMF } from '@prisma/client/runtime'

export const getPrismaClientDmmf = (packagePath: string): DMMF.Document => {
  let dmmf: undefined | DMMF.Document = undefined

  try {
    dmmf = require(packagePath).dmmf
  } catch (error) {
    throw new Error(
      `Failed to import prisma client package at ${packagePath}. The following error occured while trying:
        ${error.stack}`
    )
  }

  if (!dmmf) {
    throw new Error(`\
You most likely forgot to initialize the Prisma Client. Please run \`prisma generate\` and try to run it again.
If that does not solve your problem, please open an issue.`)
  }

  return dmmf
}
