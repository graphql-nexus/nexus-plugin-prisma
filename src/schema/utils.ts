import { core } from '@nexus/schema'
import chalk from 'chalk'
import * as fs from 'fs-jetpack'
import { GraphQLResolveInfo } from 'graphql'
import * as path from 'path'
import { inspect, isDeepStrictEqual } from 'util'

export function dump(x: any) {
  console.log(inspect(x, { depth: 20 }))
}

/**
 * Write file contents but first delete the file off disk if present. This is a
 * useful function when the effect of file delete is needed to trigger some file
 * watch/refresh mechanism, such as is the case with VSCode TS declaration files
 * inside `@types/` packages.
 *
 * For more details that motivated this utility refer to the originating issue
 * https://github.com/graphql-nexus/nexus-plugin-prisma/issues/453.
 */
export const hardWriteFile = (filePath: string, data: string): Promise<void> =>
  fs
    .removeAsync(filePath)
    .catch((error) => {
      return error.code === 'ENOENT' ? Promise.resolve() : Promise.reject(error)
    })
    .then(() => fs.writeAsync(filePath, data))

/**
 * Write file contents but first delete the file off disk if present. This is a
 * useful function when the effect of file delete is needed to trigger some file
 * watch/refresh mechanism, such as is the case with VSCode TS declaration files
 * inside `@types/` packages.
 *
 * For more details that motivated this utility refer to the originating issue
 * https://github.com/graphql-nexus/nexus-plugin-prisma/issues/453.
 */
export const hardWriteFileSync = (filePath: string, data: string): void => {
  try {
    fs.remove(filePath)
  } catch (error) {
    if (error.code !== 'ENOENT') throw error
  }
  fs.write(filePath, data)
}

// TODO `any` should be `unknown` but there is a bug (?)
// preventing that from working, see:
// https://github.com/microsoft/TypeScript/issues/33521
// https://stackoverflow.com/questions/58030413/calculate-union-type-of-key-names-in-object-whose-values-are-indexable

/**
 * TODO
 */
export const indexBy = <X extends Record<string, any>>(
  xs: X[],
  indexer: ((x: X) => string) | keyof X
): Index<X> => {
  const seed: Index<X> = {}
  if (typeof indexer === 'function') {
    return xs.reduce((index, x) => {
      const address = indexer(x)
      index[address] = x
      return index
    }, seed)
  } else {
    return xs.reduce((index, x) => {
      const address = x[indexer]
      index[address] = x
      return index
    }, seed)
  }
}

export const upperFirst = (s: string): string => {
  return s.replace(/^\w/, (c) => c.toUpperCase())
}

export function lowerFirst(s: string): string {
  if (s.length === 0) return s
  return s[0].toLowerCase() + s.slice(1)
}

export function flatMap<T, U>(array: T[], callbackfn: (value: T, index: number, array: T[]) => U[]): U[] {
  return Array.prototype.concat(...array.map(callbackfn))
}

export function dmmfFieldToNexusFieldConfig(param: {
  type: string | object
  isList: boolean
  isRequired: boolean
}): {
  type: any
  list: true | undefined
  nullable: boolean
} {
  return {
    type: param.type,
    list: param.isList ? true : undefined,
    nullable: !param.isRequired,
  }
}

export function assertPrismaClientInContext(prismaClient: any) {
  if (!prismaClient) {
    throw new Error('Could not find Prisma Client JS in context (ctx.prisma)')
  }
}

export function trimIfInNodeModules(path: string) {
  if (path.includes('node_modules')) {
    return path.substring(path.lastIndexOf('node_modules') + 'node_modules'.length + 1)
  }

  return path
}

export function getImportPathRelativeToOutput(from: string, to: string): string {
  if (to.includes('node_modules')) {
    return trimIfInNodeModules(to)
  }

  let relativePath = path.relative(from, to)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}

/**
 * Index types are just an alias for Records
 * whose keys are of type `string`. The name
 * of this type, `Index`, signifies its canonical
 * use-case for data indexed by some property, e.g.
 * a list of users indexed by email.
 */
export type Index<T> = Record<string, T>

/** TODO: Copy these types into computedInputs section of docs as part of
 * docs build process. The same should be done for other areas where code
 * has been copy pasted into our README.
 **/
/**
 *  Represents arguments required by Prisma Client JS that will
 *  be derived from a request's input (args, context, and info)
 *  and omitted from the GraphQL API. The object itself maps the
 *  names of these args to a function that takes an object representing
 *  the request's input and returns the value to pass to the prisma
 *  arg of the same name.
 */
export type LocalComputedInputs<MethodName extends string> = Record<
  string,
  (params: LocalMutationResolverParams<MethodName>) => unknown
>

export type GlobalComputedInputs = Record<string, (params: GlobalMutationResolverParams) => unknown>

type BaseMutationResolverParams = {
  info: GraphQLResolveInfo
  ctx: Context
}

export type GlobalMutationResolverParams = BaseMutationResolverParams & {
  args: Record<string, any> & { data: unknown }
}

export type LocalMutationResolverParams<MethodName extends string> = BaseMutationResolverParams & {
  args: MethodName extends keyof core.GetGen2<'argTypes', 'Mutation'>
    ? core.GetGen3<'argTypes', 'Mutation', MethodName>
    : any
}

export type Context = core.GetGen<'context'>

export const isEmptyObject = (o: any) => isDeepStrictEqual(o, {})

export function keys<A extends object>(a: A): (keyof A)[] {
  return Object.keys(a) as any
}

export function ensureDepIsInstalled(depName: string) {
  try {
    require(depName)
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      console.error(
        `${chalk.redBright(
          'ERROR:'
        )} When nexus-plugin-prisma is used as a Nexus Schema plugin, ${chalk.greenBright(
          depName
        )} must be installed as a dependency. Please run \`${chalk.greenBright(`npm install ${depName}`)}\`.`
      )
      process.exit(1)
    } else {
      throw err
    }
  }
}
