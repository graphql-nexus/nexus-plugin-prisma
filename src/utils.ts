import { relative } from 'path'
import { ExternalDMMF as DMMF } from './dmmf/dmmf-types'
import {
  OperationName,
  IFieldNamingStrategy,
} from './nexus-prisma/NamingStrategies'
import { core } from 'nexus'

export const keyBy: <T>(
  collection: T[],
  iteratee: (value: T) => string,
) => Record<string, T> = (collection, iteratee) => {
  return collection.reduce<any>((acc, curr) => {
    acc[iteratee(curr)] = curr
    return acc
  }, {})
}

export function partition<T>(
  arr: T[],
  iteratee: (val: T) => boolean,
): [T[], T[]] {
  const partitioned: [T[], T[]] = [[], []]

  for (const val of arr) {
    const partitionIndex: 0 | 1 = iteratee(val) ? 0 : 1
    partitioned[partitionIndex].push(val)
  }

  return partitioned
}

export const upperFirst = (s: string): string => {
  return s.replace(/^\w/, c => c.toUpperCase())
}

export function flatMap<T, U>(
  array: T[],
  callbackfn: (value: T, index: number, array: T[]) => U[],
): U[] {
  return Array.prototype.concat(...array.map(callbackfn))
}

export function nexusOpts(param: {
  type: core.AllInputTypes | core.AllNexusInputTypeDefs<string>
  isList: boolean
  isRequired: boolean
}): {
  type: any
  list: true | undefined
  nullable: boolean
} {
  return {
    type: param.type,
    list: param.isList ? true : undefined,
    nullable: !param.isRequired,
  }
}

export function assertPhotonInContext(photon: any) {
  if (!photon) {
    throw new Error('Could not find photon in context')
  }
}

export function trimIfInNodeModules(path: string) {
  if (path.includes('node_modules')) {
    return path.substring(
      path.lastIndexOf('node_modules') + 'node_modules'.length + 1,
    )
  }

  return path
}

export function getImportPathRelativeToOutput(
  from: string,
  to: string,
): string {
  if (to.includes('node_modules')) {
    return trimIfInNodeModules(to)
  }

  let relativePath = relative(from, to)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}

export function getCRUDFieldName(
  modelName: string,
  fieldName: string,
  mapping: DMMF.Mapping,
  namingStrategy: IFieldNamingStrategy,
) {
  const operationName = Object.keys(mapping).find(
    key => (mapping as any)[key] === fieldName,
  ) as OperationName | undefined

  if (!operationName || !namingStrategy[operationName]) {
    throw new Error(`Could not find mapping for field ${fieldName}`)
  }

  return namingStrategy[operationName](fieldName, modelName)
}
