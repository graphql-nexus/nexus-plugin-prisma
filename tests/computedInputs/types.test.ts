import { getTestData, defaultDefinitions } from '../__utils'
import { mutationType } from 'nexus'
import { ComputedInputs } from '../../src/utils'

describe('computedInputs typegen', () => {
  it('works at plugin-level', async () => {
    const { schema, typegen } = await getTestData({
      pluginOptions: {
        computedInputs: { createdWithBrowser: ({ ctx }) => ctx.browser },
      },
    })
    expect(schema).toMatchSnapshot('plugin-level-schema')
    expect(typegen).toMatchSnapshot('plugin-level-typegen')
  })
  it('works at resolver-level', async () => {
    const { schema, typegen } = await getTestData({
      definitions: {
        ...defaultDefinitions,
        mutation: mutationType({
          definition(t: any) {
            t.crud.createOneUser({
              computedInputs: {
                createdWithBrowser: ({ ctx }) => ctx.browser,
              } as ComputedInputs,
            })
            t.crud.createOneNested()
          },
        }),
      },
    })
    expect(schema).toMatchSnapshot('resolver-level-schema')
    expect(typegen).toMatchSnapshot('resolver-level-typegen')
  })
})
