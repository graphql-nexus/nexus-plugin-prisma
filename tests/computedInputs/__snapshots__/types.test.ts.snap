// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computedInputs typegen works at plugin-level: plugin-level-schema 1`] = `
"type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNested(data: NestedCreateInput!): Nested!
}

type Nested {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestedCreateInput {
  name: String!
  user: UserCreateOneWithoutNestedInput
}

input NestedCreateManyWithoutUserInput {
  create: [NestedCreateWithoutUserInput!]
  connect: [NestedWhereUniqueInput!]
}

input NestedCreateWithoutUserInput {
  name: String!
}

input NestedWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
}

type User {
  id: Int!
  name: String!
  nested(skip: Int, after: NestedWhereUniqueInput, before: NestedWhereUniqueInput, first: Int, last: Int): [Nested!]!
  createdWithBrowser: String!
}

input UserCreateInput {
  name: String!
  nested: NestedCreateManyWithoutUserInput
}

input UserCreateOneWithoutNestedInput {
  create: UserCreateWithoutNestedInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestedInput {
  name: String!
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`computedInputs typegen works at plugin-level: plugin-level-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nested: prisma.Nested
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'name' | 'nested' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}
    nesteds: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nested: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nested: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nested: 'Nested'
    nesteds: 'Nested'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNested: 'Nested'
    updateOneNested: 'Nested'
    updateManyNested: 'BatchPayload'
    deleteOneNested: 'Nested'
    deleteManyNested: 'BatchPayload'
    upsertOneNested: 'Nested'

  },
  User: {
    id: 'Int'
    name: 'String'
    nested: 'Nested'
    createdWithBrowser: 'String'

},  Nested: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nested: NexusPrismaFields<'Nested'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`computedInputs typegen works at resolver-level: resolver-level-schema 1`] = `
"type Mutation {
  createOneUser(data: UserCreateWithoutCreatedWithBrowserInput!): User!
  createOneNested(data: NestedCreateInput!): Nested!
}

type Nested {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestedCreateInput {
  name: String!
  createdWithBrowser: String!
  user: UserCreateOneWithoutNestedInput
}

input NestedCreateManyWithoutUserWithoutCreatedWithBrowserInput {
  create: [NestedCreateWithoutUserWithoutCreatedWithBrowserInput!]
  connect: [NestedWhereUniqueWithoutCreatedWithBrowserInput!]
}

input NestedCreateWithoutUserWithoutCreatedWithBrowserInput {
  name: String!
}

input NestedWhereUniqueInput {
  id: Int
}

input NestedWhereUniqueWithoutCreatedWithBrowserInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
}

type User {
  id: Int!
  name: String!
  nested(skip: Int, after: NestedWhereUniqueInput, before: NestedWhereUniqueInput, first: Int, last: Int): [Nested!]!
  createdWithBrowser: String!
}

input UserCreateOneWithoutNestedInput {
  create: UserCreateWithoutNestedInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutCreatedWithBrowserInput {
  name: String!
  nested: NestedCreateManyWithoutUserWithoutCreatedWithBrowserInput
}

input UserCreateWithoutNestedInput {
  name: String!
  createdWithBrowser: String!
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`computedInputs typegen works at resolver-level: resolver-level-typegen 1`] = `
"
interface ModelTypes {
  User: prisma.User
  Nested: prisma.Nested
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'name' | 'nested' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}
    nesteds: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {
    nested: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },  Nested: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nested: 'Nested'
    nesteds: 'Nested'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNested: 'Nested'
    updateOneNested: 'Nested'
    updateManyNested: 'BatchPayload'
    deleteOneNested: 'Nested'
    deleteManyNested: 'BatchPayload'
    upsertOneNested: 'Nested'

  },
  User: {
    id: 'Int'
    name: 'String'
    nested: 'Nested'
    createdWithBrowser: 'String'

},  Nested: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nested: NexusPrismaFields<'Nested'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;
